File: backend-dotnet/Application/Queries/Posts/GetPostByIdQueryHandler.cs
using Application.DTOs;
using Application.Interfaces;
using Domain.Entities;
using Domain.Enums;
using Microsoft.Extensions.Logging;

namespace Application.Queries.Posts
{
    public class GetPostByIdQueryHandler : IQueryHandler<GetPostByIdQuery, PostListItemDto>
    {
        private readonly IPostRepository _postRepository;
        private readonly ICacheService _cacheService;
        private readonly ILogger<GetPostByIdQueryHandler> _logger;

        public GetPostByIdQueryHandler(
            IPostRepository postRepository,
            ICacheService cacheService,
            ILogger<GetPostByIdQueryHandler> logger)
        {
            _postRepository = postRepository;
            _cacheService = cacheService;
            _logger = logger;
        }

        public async Task<PostListItemDto> HandleAsync(GetPostByIdQuery query)
        {
            var cacheKey = $"post_details_{query.PostId}_{query.CurrentUserId}";
            
            var cached = await _cacheService.GetAsync<PostListItemDto>(cacheKey);
            if (cached != null)
            {
                _logger.LogDebug("Post {PostId} retrieved from cache", query.PostId);
                return cached;
            }

            // Check if post exists and get visibility permissions in one call
            var result = await _postRepository.GetPostDtoAsync(query.PostId, query.CurrentUserId);
            
            if (result == null)
            {
                throw new KeyNotFoundException("Post not found");
            }

            // Get the post entity to check visibility rules
            var post = await _postRepository.GetPostWithUserAndLikesAsync(query.PostId);
            
            if (post == null)
            {
                throw new KeyNotFoundException("Post not found");
            }

            // Check visibility permissions
            if (post.Visibility == PostVisibility.Private && post.UserId != query.CurrentUserId)
            {
                throw new UnauthorizedAccessException("You don't have permission to view this post");
            }

            if (post.Visibility == PostVisibility.FollowersOnly && query.CurrentUserId.HasValue && post.UserId != query.CurrentUserId)
            {
                var isFollowing = await _postRepository.IsFollowerAsync(query.CurrentUserId.Value, post.UserId);
                
                if (!isFollowing)
                {
                    throw new UnauthorizedAccessException("You don't have permission to view this post");
                }
            }

            // Cache for 30 minutes
            await _cacheService.SetAsync(cacheKey, result, TimeSpan.FromMinutes(30));

            _logger.LogInformation("Post {PostId} retrieved successfully", query.PostId);
            return result;
        }
    }
}

File: backend-java/src/main/java/com/jerrygram/application/queries/posts/GetPostByIdQueryHandler.java
package com.jerrygram.application.queries.posts;

import com.jerrygram.application.dtos.PostListItemDto;
import com.jerrygram.application.interfaces.ICacheService;
import com.jerrygram.application.interfaces.IQueryHandler;
import com.jerrygram.domain.enums.PostVisibility;
import com.jerrygram.infrastructure.repositories.PostRepository;
import com.jerrygram.infrastructure.repositories.UserFollowRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.NoSuchElementException;

@Service
@RequiredArgsConstructor
@Slf4j
public class GetPostByIdQueryHandler implements IQueryHandler<GetPostByIdQuery, PostListItemDto> {

    private final PostRepository postRepository;
    private final ICacheService cacheService;
    private final UserFollowRepository userFollowRepository;

    @Override
    public PostListItemDto handle(GetPostByIdQuery query) {
        var postId = query.getPostId();
        var currentUserId = query.getCurrentUserId();
        
        var cacheKey = "post_details_" + postId + "_" + currentUserId;
        
        var cached = cacheService.get(cacheKey, PostListItemDto.class);
        if (cached.isPresent()) {
            log.debug("Post {} retrieved from cache", postId);
            return cached.get();
        }

        // Get the post entity to check visibility rules first
        var post = postRepository.getPostWithUserAndLikes(postId);
        
        if (post == null) {
            throw new NoSuchElementException("Post not found");
        }

        // Check visibility permissions
        if (post.getVisibility() == PostVisibility.Private && !post.getUserId().equals(currentUserId)) {
            throw new SecurityException("You don't have permission to view this post");
        }

        if (post.getVisibility() == PostVisibility.FollowersOnly && currentUserId != null && !post.getUserId().equals(currentUserId)) {
            var isFollowing = userFollowRepository.existsByFollowerIdAndFollowingId(currentUserId, post.getUserId());
            
            if (!isFollowing) {
                throw new SecurityException("You don't have permission to view this post");
            }
        }

        // Get the DTO after permission checks pass
        var result = postRepository.getPostDto(postId, currentUserId);
        
        if (result == null) {
            throw new NoSuchElementException("Post not found");
        }

        // Cache for 30 minutes
        cacheService.set(cacheKey, result, Duration.ofMinutes(30));

        log.info("Post {} retrieved successfully", postId);
        return result;
    }
}